### PR TITLE
Remove torch_scatter and torch_cluster from CI pipeline's dependencies

### PR DESCRIPTION
**Pull request by dhpitt**
We currently use the module `torch_scatter` on main for one function called `segment_csr`, which aggregates features across each neighborhood for the GINO model in `neuralop.layers.integral_transform.py`. This implements a simple python version of the same function to remove our dependence on `torch_scatter` and fix that annoying OSError in our CI.

**Second note**: the same error occurs with the CPU version of `torch_cluster`, so I wrote a neighborhood search for CPU runners only.

### CODE REVIEW

**Comment by dhpitt** [comment]
Small empirical tests show that my neighbor search function runs in about 1/3 the time of `torch_cluster.radius`.

File: neuralop/layers/neighbor_search.py [hunk 1]
```diff
@@ -61,31 +69,34 @@ def forward(self, data, queries, radius):
...
-                bins[-1] -= 1
-            
-            neighbors_row_splits = torch.cumsum(bins, dim=0)
-            neighbors_row_splits = torch.cat((torch.tensor([0], 
-                                              dtype=neighbors_row_splits.dtype,
-                                              device=neighbors_row_splits.device),
-                                              neighbors_row_splits), dim=0)
-            
+                neighbors_count, neighbors_index = self.search_fn(data, queries, radius,
+                                                max_num_neighbors=data.shape[0])
```
Comment (JeanKossaifi) [inline neuralop/layers/neighbor_search.py:76]: Should we wrap that code blow in a small, tested function?

File: neuralop/layers/neighbor_search.py [hunk 2]
```diff
@@ -61,31 +69,34 @@ def forward(self, data, queries, radius):
...
-            
-            if neighbors_count[-1] != queries.shape[0] - 1:
-                add_max_element = True
-                neighbors_count = torch.cat((neighbors_count,
-                                             torch.tensor([queries.shape[0] - 1], 
-                                                          dtype=neighbors_count.dtype,
-                                                          device=neighbors_count.device)), 
-                                                          dim=0)
+            if not self.use_torch_cluster:
+                return_dict = self.search_fn(data, queries, radius)
```
Comment (JeanKossaifi) [inline neuralop/layers/neighbor_search.py:73]: @kovachki Should we just keep the manual implementation and just choose between open3d (fast) and fallback (slow)?

File: neuralop/layers/segment_csr.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        reduce by taking the average of all neighbors.
+        Otherwise take the sum. 
+    """
+    if reduce not in ['mean', 'sum']:
+        raise ValueError("reduce must be one of \'mean\', \'sum\'")
+    
+    if torch.backends.cuda.is_built():
+        """only import torch_scatter when cuda is available"""
+        import torch_scatter.segment_csr as scatter_segment_csr
+        return scatter_segment_csr(src, indptr, reduce)
```
Comment (JeanKossaifi) [inline neuralop/layers/segment_csr.py:33]: Hmm cuda being built doesn't necessarily mean torch_scatter will be available? Shouldn't we check for both?

**Review by JeanKossaifi — COMMENTED** [review]
Thanks for the changes @dhpitt  -- should we have a small unit-test for the fallbacks?

File: neuralop/layers/neighbor_search.py (hunk 1 above)
Comment (dhpitt) [reply neuralop/layers/neighbor_search.py:76]: Added a unit test for fallback search

Comment (dhpitt) [reply neuralop/layers/neighbor_search.py:76]: Torch_cluster is slower than my implementation anyway. What if we just axed that coed

File: neuralop/layers/neighbor_search.py (hunk 2 above)
Comment (dhpitt) [reply neuralop/layers/neighbor_search.py]: Implemented.

File: neuralop/layers/segment_csr.py (hunk 3 above)
Comment (dhpitt) [reply neuralop/layers/segment_csr.py:31]: Logic is changed.

File: neuralop/layers/neighbor_search.py (hunk 2 above)
Comment (kovachki) [reply neuralop/layers/neighbor_search.py]: Yes, I think this is best. Let's just remove cluster altogether.

File: neuralop/layers/segment_csr.py (hunk 3 above)
Comment (kovachki) [reply neuralop/layers/segment_csr.py:31]: @dhpitt could you please check that your segment_csr works with backwards. You're doing some inplace operations so I'm not sure.

**Comment by JeanKossaifi** [comment]
Awesome, thanks @dhpitt !

**Pull request merged**